File: PYTHON/5-BinarySearch/3-min_eating_speed.py
from typing import List
import math

# brute force
# def minEatingSpeed(piles: List[int], h: int) -> int:
#     speed = 1
#     while True:
#         totalTime = 0
#         for pile in piles:
#             totalTime += math.ceil(pile / speed)
#         if totalTime <= h:
#             return speed
#         speed += 1

# binary search
def minEatingSpeed(piles: List[int], h: int) -> int:
    left, right = 1, max(piles)
    res = right

    while left <= right:
        k = (left + right) // 2
        totalTime = 0
        for pile in piles:
            totalTime += math.ceil(float(pile) / k)
        if totalTime <= h:
            res = k
            right = k - 1
        else:
            left = k + 1
    
    return res

if __name__ == "__main__":
    piles = [25,10,23,4]
    h = 4
    print(minEatingSpeed(piles, h))

File: CPP/5-BinarySearch/3-min_eating_speed.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// brute force
int minEatingSpeed(vector<int>& piles, int h) {
    int speed = 1;
    while (true) {
        long long totalTime = 0;
        for (int pile : piles) {
            totalTime += (pile + speed - 1) / speed;
        }

        if (totalTime <= h) {
            return speed;
        }

        speed++;
    }
}

int main() {
    vector<int> piles = {1, 4, 3, 2};
    int h = 9;
    cout << minEatingSpeed(piles, h);
}